# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(50)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   
20         1   -1311.604166         36              среднее             1   
21         1    -253.685166         33              среднее             1   
22         1   -1766.644138         24              среднее             1   
23         0    -272.981385         21               высшее             0   
24         1  338551.952911         57              среднее             1   
25         0  363548.489348         67              среднее             1   
26         0            NaN         41              среднее             1   
27         0    -529.191635         28               высшее             0   
28         1    -717.274324         26               высшее             0   
29         0            NaN         63              среднее             1   
30         1  335581.668515         62              среднее             1   
31         0   -1682.083438         47            начальное             3   
32         0   -4649.910832         34              СРЕДНЕЕ             1   
33         0   -1548.637544         48              среднее             1   
34         0   -4488.067031         35              среднее             1   
35         0  394021.072184         68              среднее             1   
36         0    -176.216688         33              среднее             1   
37         0   -6448.810860         43               ВЫСШЕЕ             0   
38         0    -597.881827         25               высшее             0   
39         0    -650.587796         31               высшее             0   
40         0   -1030.362606         30              среднее             1   
41         0            NaN         50              среднее             1   
42         0   -1257.496190         20  неоконченное высшее             2   
43         0   -4375.681384         43  неоконченное высшее             2   
44         1   -1362.041728         26              СРЕДНЕЕ             1   
45         1   -1039.451741         49              среднее             1   
46         2   -2262.712304         37              среднее             1   
47         1   -2689.137274         33               высшее             0   
48         0   -3341.067886         45              среднее             1   
49         0   -1181.443228         54              среднее             1   

            family_status  family_status_id gender  income_type  debt  \
0  

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Функция вернет значение по аргументу, в котором посчитает количество кредитов, сумму и среднее значение.
А так же перезапишет индесы. 

In [22]:
def data_pivot_column (column):
    result = data.pivot_table(index=column, values='debt', aggfunc=['count', 'sum', 'mean'])\
    .reset_index(level=column)
    return result 

In [23]:
child_quantity = data_pivot_column('children')

In [24]:
child_quantity.columns = ['кол-во детей', 'всего_кредитов', 'всего_должников', 'процент_должников']

In [25]:
child_quantity['процент_должников'] = [f'{x:,.2%}' for x in child_quantity['процент_должников']]

In [26]:
child_quantity.sort_values(by='процент_должников')

кол-во детей  всего_кредитов  всего_должников процент_должников
5             5               9                0             0.00%
0             0           14091             1063             7.54%
3             3             330               27             8.18%
1             1            4808              444             9.23%
2             2            2052              194             9.45%
4             4              41                4             9.76%

**Вывод:** Бездетные - самая большая категория из тех, кто обращался за кредитом.  В этой группе самая низкая вероятность не возврата кредита в срок, всего 7.5%.  Заемщики, имеющие от одного до двух детей, обращаются за кредитом в два раза меньше от общего числа и рост невозврата кредита в срок в этой категории  почти 9.5%. Многодетные – самая малочисленная категория из всех,  сложно сделать более точные прогнозы по этой группе, но можно предположить,  что риск не возврата будет выше, чем у первой группы заемщиков. **Так что, закономерность имеет место быть, наличие детей увеличивает риск невозврата кредита в срок.**  

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [27]:
status_family = data_pivot_column('family_status')

In [28]:
status_family.columns = ['семейное положение', 'всего_кредитов', 'всего_должников', 'процент_должников']

In [29]:
status_family['процент_должников'] = [f'{x:,.2%}' for x in status_family['процент_должников']]

In [30]:
status_family.sort_values(by='процент_должников')

семейное положение  всего_кредитов  всего_должников процент_должников
2         вдовец / вдова             951               63             6.62%
1              в разводе            1189               84             7.06%
4        женат / замужем           12261              927             7.56%
3       гражданский брак            4134              385             9.31%
0  Не женат / не замужем            2796              273             9.76%

**Вывод:** Чаще всего обращаются за кредитом люди, состоящие в браке, процент не возвращенных кредитов в срок составляет 7.5%. В то время как те, кто не состоит в браке или живет гражданским браком прибегает к кредиту не часто, но риск не возврата более 9%. Овдовевшие и разведенные самая малочисленная категория потребителей кредита с риском не возврата  районе 7 %. **Здесь тоже прослеживается закономерность,  люди, бывшие в браке или состоящие в нем, более ответственно подходят к исполнению своих обязательств по возврату кредита в срок.** 

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [31]:
income_total = data_pivot_column('total_income_category')

In [32]:
income_total.columns = ['уровень дохода', 'всего_кредитов', 'всего_должников', 'процент_должников']

In [33]:
income_total['процент_должников'] = [f'{x:,.1%}' for x in income_total['процент_должников']]

In [34]:
income_total.sort_values(by='процент_должников')


уровень дохода  всего_кредитов  всего_должников процент_должников
3              D             349               21              6.0%
1              B            5014              354              7.1%
0              A              25                2              8.0%
2              C           15921             1353              8.5%
4              E              22                2              9.1%

**Медианный доход по каждой категории**

In [35]:
data.groupby('total_income_category')['total_income'].agg('median')

total_income_category
A    1223042
B     257558
C     130355
D      43929
E      26431
Name: total_income, dtype: int64

**Вывод:** Здесь выделяется основная группа с доходом 130 000Р на нее приходится 75% всех выданных кредитов. В ней 8.5 % должников.  Люди с доходом до 30 000Р редко берут кредит и процент невозврата более 9% Люди с доходом 43 000Р Чаще всех во время отдают кредиты 6% должников. **Можно сказать, что чем больше доход, тем больше вероятность возврата кредита в срок.**

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [36]:
purpose_credit = data_pivot_column('purpose_category')

In [37]:
purpose_credit.columns = ['цель кредита', 'всего_кредитов', 'всего_должников', 'процент_должников']

In [38]:
purpose_credit['процент_должников'] = [f'{x:,.2%}' for x in purpose_credit['процент_должников']]

In [39]:
purpose_credit.sort_values(by='процент_должников')

цель кредита  всего_кредитов  всего_должников процент_должников
1  операции с недвижимостью           10751              780             7.26%
3        проведение свадьбы            2313              183             7.91%
2     получение образования            3988              369             9.25%
0    операции с автомобилем            4279              400             9.35%

**Вывод:** Половина всех кредитов это кредиты на операции с недвижимостью, процент должников 7.2% почти столько же задолженностей по кредитам на проведение свадьбы, а вот кредиты на образование и покупку авто имеют должников более  9% . **Кредиты на операции с недвижимостью чаще возвращают в срок.** 

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* ***В наших данных пропуски в столбцах "трудовой стаж" и "ежемесячный доход" абсолютно идентичны. Количество строк с отсутствующими данными в обоих указанных столбцах совпадает и равно 2174. Маловероятно, что специалисты по выдаче кредитов, заполнив все без исключения ячейки других столбцов по каждому кредитополучателю, оставляли незаполненными именно два показателя, и сделали это ровно 2174 раза в каждом столбце. Даже не проводя проверку, во всех ли случаях незаполненные ячейки находятся в одной и то же строке с высокой долей вероятности можно утверждать, что произошла техническая ошибка при формировании таблицы с исходными данными (либо некий сбой во время выгрузки данных на анализ, либо при объединении данных в единый файл, который также должен был передаваться на анализ, либо эти данные не собирались).

Пропуски в столбцах 'total_income' и 'days_employed' соответствуют значениям 'пенсионер', 'госслужащий', 'компаньон', 'сотрудник' и 'предприниматель' в столбце 'income_type'. Однако в 'income_type' также присутствуют значения ‘студент’, ‘безработный’ и ‘в декрете’, в которых пропусков нет вообще. На основании этого можно предположить, что исходная таблица была объединена с другим датасетом, содержавшим сведения о категориях 'пенсионер', 'госслужащий', 'компаньон', 'сотрудник' и 'предприниматель', но без сведений о доходе и трудовом стаже. В результате при объединении таблиц столбцы 'total_income' и 'days_employed' были заполнены значением NaN..*** 

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* ***В количественных переменных, среднее значение не корректно характеризует данные, так как некоторые данные могут сильно выделяться из общего набора. Лучше исключить их при выборке, а взять медианное значение для более точного прогноза.***  

### Шаг 4: общий вывод.


    **Было проведено исследование, что бы выявить влияние разных факторов на погашение кредита  срок.  Исходные данные – статистическая таблица  платежеспособности клиентов. Результаты исследования будут учтены при построении модели  кредитного скоринга – специальной системы, которая оценивает способность потенциального заемщика вернуть кредит банку.**

*Перед проведением исследования были поставлены несколько гипотез.*

**Гипотеза 1:** 
Есть ли зависимость между количеством детей и возвратом кредита в срок?
**Гипотеза 2:** 
Есть ли зависимость между семейным положением и возвратом кредита в срок?
**Гипотеза 3:** 
Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
**Гипотеза 4:** 
Как разные цели кредита влияют на его возврат в срок?

*В результате исследования удалось выявить зависимости*

**Гипотеза 1** частично подтверждена.  Для более точных выводов необходимо больше данных по клиентам с 3, 4 и 5 детьми.

7.54%  просроченных кредитов у заемщиков без детей.
9.23%  просроченных кредитов у заемщиков с 1 ребенком.
9.45 % просроченных кредитов у заемщиков с 2-я детьми.

***Рекомендация: провести дополнительные исследования на большей выборке данных.***
    
**Гипотеза 2** подтверждена.
    
6.62% - 7.56%  просроченных кредитов у заемщиков имеющих семейный статус, вдовец / вдова, в разводе, женат / замужем. Вероятность возврата кредита больше.
9.31% - 9.76% просроченных кредитов у заемщиков, не состоящих в браке или живущих гражданским браком.  Вероятность возврата кредита меньше. 
 
***Рекомендация: Так как разница в данных у этих групп не значительная не стоит уделять основное внимание  данной зависимости при принятии решения о платежеспособности  клиента.***    
**Гипотеза 3** Частично подтверждена для более точных выводов нужно больше данных для клиентов с доходом в категории А (более 1 000 000Р)  D (от 30 001Р- 50 000Р) Е (до 30 000Р)
    
6% просроченных кредитов у заемщиков с доходом 43 929 руб.
7.1% просроченных кредитов у заемщиков с доходом 257 558 руб.
8% просроченных кредитов у заемщиков с доходом 1 223 042 руб. 
8.5% просроченных кредитов у заемщиков с доходом 130 355 руб.
9.1% просроченных кредитов у заемщиков с доходом 26 431 руб.
    
***Рекомендация: Провести дополнительные исследования на большей выборке данных***
    
**Гипотеза 4** подтверждена.
    
7.26% просроченных кредитов на операции с надвижимостью 
7.91% просроченных кредитов на проведение свадьбы
9.25% просроченных кредитов на получение образования
9.35% просроченных кредитов на операции с автомобилем
    
***Рекомендации: Не стоит уделять решающее значение при принятии решения о платежеспособности клиента по разнице в этих категориях.*** 

<b>Общие рекомендации<b>
    
***Специалистам при построении кредитного скоринга можно использовать зависимости гипотез 2 и 4. (Семейное положение и цель кредита)

    Необходимо собрать больше данных для повторной проверки Гипотез 1 и 3.***